qark/qarkMain.py

```diff
@@ -214,12 +214,19 @@ def process_manifest(manifest):
                 exit()
             common.logger.error(str(e) + "\r\nThat didnt work. Try providing an absolute path to the file\n")
 
+def hasmode(filename, mode):
+    st = os.stat(filename)
+    return (st.st_mode & mode)
+
+def setmode(filename, mode):
+    st = os.stat(filename)
+    os.chmod(filneame, st.st_mode | mode)
 
 def list_all_apk():
     result = []
     adb = common.getConfig('AndroidSDKPath') + "platform-tools/adb"
     st = os.stat(adb)
-    os.chmod(adb, st.st_mode | stat.S_IEXEC)
+    if not hasmode(adb, stat.S_IEXEC): setmode(adb, stat.S_IEXEC)
```

Comment:
While this code implements the required logic, it is an example of a style of programming that Python doesn't encourage and doesn't optimize for. Specifically, this is referred to as "Look Before You Leap" where Python generally prefers "Easier to ask for forgiveness than permission" [EAFP](https://docs.python.org/3.4/glossary.html?highlight=eafp).

If you wanted to refactor you could catch the exception that you saw earlier.

```
try:
    os.chmod(adb, st.st_mode | stat.S_IEXEC)
except OSError:
    try:
        hasmode(adb, stat.S_IEXEC)
        common.logger.info('adb is already executable')
    except Exception as ex:
        common.logger.error('Exception trying to stat adb: {}'.format(ex))
except Exception as ex:
    common.logger.error('Exception trying to chmod adb: {}'.format(ex))
```

That said, there are plenty of instances of non-pythonic code patterns still in Qark 😄